modules/core/src/norm.cpp

```diff
@@ -205,6 +205,773 @@ int normL1_(const uchar* a, const uchar* b, int n)
+                res = std::max(tres, res);
+#else
+                tres = std::max(v_reduce_max(v_int32(max0, max1, max2, max3)), v_reduce_max(v_int32(max4, max5, max6, max7)));
+                res = std::max(tres, res);
+#endif
+            }
+
+            a += ( i * cn );
+            b += ( i * cn );
+#endif
+
+            for(; i < len; i++, a += cn, b += cn )
+            {
+                if( mask[i] )
+                {
+                    for(int j = 0; j < cn; j++ )
+                        res = std::max(res, (int)std::abs(a[j] - b[j]));
+                }
+            }
+        } else if ( cn == 3 )
+        {
+#if CV_SIMD
+            const v_uint32 zero = vx_setall_u32((int)0);
+            for(; i*cn <= len - v_uint8::nlanes*cn; i += v_uint8::nlanes )
+            {
+                v_uint8 m = vx_load(mask + i);
+
+                v_uint16 m0, m1;
+
+                v_expand(m, m0, m1);
+
+                v_uint32 m00, m01, m10, m11;
+
+                v_expand(m0, m00, m01);
+                v_expand(m1, m10, m11);
+
+                m00 = m00 != zero;
+                m01 = m01 != zero;
+                m10 = m10 != zero;
+                m11 = m11 != zero;
+
+                v_int32 v_res = vx_setall_s32(res);
+
+                v_uint8 v_srca0, v_srca1, v_srca2;
+                v_load_deinterleave(a + i * cn, v_srca0, v_srca1, v_srca2);
+                v_uint16 v_srca00, v_srca01, v_srca10, v_srca11;
+                v_uint16 v_srca20, v_srca21;
+
+                v_expand(v_srca0, v_srca00, v_srca01);
+                v_expand(v_srca1, v_srca10, v_srca11);
+                v_expand(v_srca2, v_srca20, v_srca21);
+
+                v_uint32 v_srca000, v_srca001, v_srca010, v_srca011;
+                v_uint32 v_srca100, v_srca101, v_srca110, v_srca111;
+                v_uint32 v_srca200, v_srca201, v_srca210, v_srca211;
+
+                v_expand(v_srca00, v_srca000, v_srca001);
+                v_expand(v_srca01, v_srca010, v_srca011);
+                v_expand(v_srca10, v_srca100, v_srca101);
+                v_expand(v_srca11, v_srca110, v_srca111);
+                v_expand(v_srca20, v_srca200, v_srca201);
+                v_expand(v_srca21, v_srca210, v_srca211);
+
+                v_uint8 v_srcb0, v_srcb1, v_srcb2;
+                v_load_deinterleave(b + i * cn, v_srcb0, v_srcb1, v_srcb2);
+                v_uint16 v_srcb00, v_srcb01, v_srcb10, v_srcb11;
+                v_uint16 v_srcb20, v_srcb21;
+
+                v_expand(v_srcb0, v_srcb00, v_srcb01);
+                v_expand(v_srcb1, v_srcb10, v_srcb11);
+                v_expand(v_srcb2, v_srcb20, v_srcb21);
+
+                v_uint32 v_srcb000, v_srcb001, v_srcb010, v_srcb011;
+                v_uint32 v_srcb100, v_srcb101, v_srcb110, v_srcb111;
+                v_uint32 v_srcb200, v_srcb201, v_srcb210, v_srcb211;
+
+                v_expand(v_srcb00, v_srcb000, v_srcb001);
```

Comment:
I agree although I don't know a much better way to implement this than just use what I'm already doing (maybe that's what you wanted?) and if we're going to implement v_select3 maybe 2 and 4 channels should be handled as well?

Reply:
`v_select3()` approach should introduce `v_select2` and `v_select4` too.

Reply:
What about something like 
```
template<typename _Tp, int n> inline v_reg<_Tp, n> v_select3(const v_reg<_Tp, n>& mask,
                                                           const v_reg<_Tp, n>& a, const v_reg<_Tp, n>& b);
template<typename _Tp, int n> inline v_reg<_Tp, n> v_select2(const v_reg<_Tp, n>& mask,
                                                           const v_reg<_Tp, n>& a, const v_reg<_Tp, n>& b);
template<typename _Tp, int n> inline v_reg<_Tp, n> v_select4(const v_reg<_Tp, n>& mask,
                                                           const v_reg<_Tp, n>& a, const v_reg<_Tp, n>& b);
```
although I think adding the number of channels as a template argument to v_select might also be an interesting path since C++17 supports default template arguments nothing would break.